applehat/src/ui.rs

```diff
@@ -0,0 +1,285 @@
+                    } else {
+                        None
+                    },
+                );
+            } else {
+                self.alphanum.print_str("gone", false);
+            }
+        } else {
+            self.alphanum.print_str("    ", false);
+        }
+        if let Err(e) = self.alphanum.show() {
+            error!("Error displaying: {}", e);
+        }
+    }
+
+    fn button_pressed(&mut self, button_index: usize) {
+        debug!("Button {} pressed.", button_index);
+        match button_index {
+            0 => {
+                // Select next node.
+                let devices = self.controller.devices();
+                let nodes = find_nodes(&devices);
+                if !nodes.is_empty() {
+                    let new_index = if let Some(selected_node_id) = &self.selected_node_id {
+                        if let Some(current_index) = nodes
+                            .iter()
+                            .position(|(_, node_id, _)| node_id == selected_node_id)
+                        {
+                            (current_index + 1) % nodes.len()
+                        } else {
+                            0
+                        }
+                    } else {
+                        0
+                    };
+                    self.selected_node_id = Some(nodes[new_index].1.to_string());
+                }
+            }
+            1 => {
+                // Select next property.
+                let current_index = PROPERTY_IDS
+                    .iter()
+                    .position(|x| x == &self.selected_property_id)
+                    .unwrap_or(0);
+                self.selected_property_id =
+                    PROPERTY_IDS[(current_index + 1) % PROPERTY_IDS.len()].to_string();
+            }
+            2 => {
+                // Change brightness.
+                let current_index = BRIGHTNESS_LEVELS
+                    .iter()
+                    .position(|&level| level == self.selected_brightness)
+                    .unwrap_or(0);
+                self.selected_brightness =
+                    BRIGHTNESS_LEVELS[(current_index + 1) % BRIGHTNESS_LEVELS.len()];
+            }
+            _ => {}
+        }
+        self.update_display();
+    }
+
+    fn update_button_state(&mut self, new_state: [bool; 3]) {
+        for i in 0..3 {
+            if new_state[i] && !self.button_state[i] {
+                self.button_pressed(i);
+            }
+        }
+        self.button_state = new_state;
+    }
+}
+
+pub fn spawn_button_poll_loop(
+    mut buttons: Buttons,
+    ui_state: Arc<Mutex<UiState>>,
+) -> JoinHandle<()> {
+    task::spawn(async move {
+        loop {
+            let new_state = [
+                buttons.a.is_pressed(),
+                buttons.b.is_pressed(),
+                buttons.c.is_pressed(),
+            ];
+            ui_state.lock().unwrap().update_button_state(new_state);
+
+            sleep(BUTTON_POLL_PERIOD).await;
+        }
```

Comment:
:'( killing the planet 100ms at a time. I guess there is no way to wait for a button click on an event driven way?

Does this put you to of the leaderboard on powertop or is there something even more sad running there already?

Reply:
Not with the library I'm using. It looks like the underlying hardware does support interrupts though, so maybe I can see about adding support for that.

I haven't checked powertop yet, I'll have a look tonight.

Reply:
It comes in between 3rd and 8th place, e.g.:

```
Power est.              Usage       Events/s    Category       Description
  3.94 W      3.5 ms/s     995.8        Interrupt      [51] dwc_otg_sim-fiq
  141 mW    250.0 µs/s      35.7        Interrupt      [74] dwc_otg
 28.7 mW      0.9 ms/s       6.9        Process        [PID 640] /usr/bin/telegraf -config /etc/telegraf/telegraf.conf -config-directory /etc/telegraf/telegraf.d
 27.5 mW    148.4 µs/s       6.9        Interrupt      [6] IPI
 25.2 mW      0.8 ms/s       6.1        Timer          hrtimer_wakeup
 25.1 mW    578.4 µs/s       6.1        Timer          tick_sched_timer
 24.0 mW      1.2 ms/s       5.6        Interrupt      [6] tasklet(softirq)
 17.2 mW      2.9 ms/s       3.2        Process        [PID 886] /home/pi/src/mijia-homie/target/debug/applehat
```